### PR TITLE
Communication disconnection timeout implementation

### PR DESCRIPTION
**Pull request by AntumArk**
I have made that when you disconnect your device from hoverboard, it could either slow down or make hard stop.
Soft stop is enabled by default. Timeout value is configurable, as well as enabling HARD stop. Everything can be changed in either config.h or platformio.ini

This feature is used with Position/Speed/PWM controls and with UART2/UART3/SOft communications. 

Here is demonstration video: 
  https://photos.app.goo.gl/V6zawA2GK3L7XRNs5

### CODE REVIEW

File: inc/config.h [hunk 1]
```diff
@@ -101,6 +101,13 @@
     //#define DEBUG_SERIAL_ASCII
   #endif
 
+// ############################### INPUT ###############################
```
Comment (p-h-a-i-l) [inline inc/config.h]: there is another section called input in line 285

File: inc/config.h [hunk 2]
```diff
@@ -101,6 +101,13 @@
...
+  #ifndef INPUT_TIMEOUT
+    #define INPUT_TIMEOUT          200          // time (ms) when to set motor enable=0, if there is no motor input
```
Comment (p-h-a-i-l) [inline inc/config.h]: lines 188 ff should be removed when you rename the variable

File: inc/config.h [hunk 3]
```diff
@@ -101,6 +101,13 @@
...
+  #endif
+//#ifndef HARD_STOP
+  //#define HARD_STOP                    // choice to stop right away or to decrement to zero
```
Comment (p-h-a-i-l) [inline inc/config.h]: rename to something which can be easily associated with the INPUT_TIMEOUT

File: src/main.c [hunk 4]
```diff
@@ -101,6 +101,10 @@ extern float batteryVoltage; // global variable for battery voltage
 uint32_t inactivity_timeout_counter;
 uint32_t debug_counter = 0;
 
+#ifdef INPUT_TIMEOUT
+  volatile uint32_t input_timeout;
```
Comment (p-h-a-i-l) [inline src/main.c]: also remove int timeout

File: src/main.c [hunk 5]
```diff
@@ -469,6 +473,51 @@ int main(void) {
...
+            input_timeout=0;
+        break;
+      case CONTROL_TYPE_PWM:
+            #ifdef HARD_STOP
+                    pwms[0] = 0;
+                    pwms[1] = 0;
+            #else
+                    pwms[0] = 0;
+                    pwms[1] = 0;
+                    enable = 0;
```
Comment (p-h-a-i-l) [inline src/main.c]: whats the purpose of 
```
                    pwms[0] = 0;
                    pwms[1] = 0;
```
?
pulling enable should ignore all inputs anyway.

Also: How do you recover enable=1 when communication continues?

**Review by p-h-a-i-l — COMMENTED** [review]
Som remarks, also we should make this a general solution which also works with PPM and I2C

File: src/main.c (hunk 5 above)
Comment (AntumArk) [reply src/main.c]: You are correct, `pwms[0] = 0;
                    pwms[1] = 0; ` is not needed, as the same thing is done at line 814 at main.c
Inside protocol.c, when write function is called.
https://github.com/bipropellant/bipropellant-protocol/pull/28/commits/47184890113e3842b308228e231991ca589b1c44#diff-6da6886a4c67f48a449b698adc074a05R222

File: inc/config.h (hunk 1 above)
Comment (AntumArk) [reply inc/config.h]: Changed to INPUT_TIMEOUT. Is it ok?

File: inc/config.h (hunk 2 above)
Comment (AntumArk) [reply inc/config.h]: You mean     -D IGNORE_GLOBAL_CONFIG ?

File: inc/config.h (hunk 3 above)
Comment (AntumArk) [reply inc/config.h]: How about `INPUT_TIMEOUT_USE_HARD_STOP`?

File: src/main.c (hunk 4 above)
Comment (AntumArk) [reply src/main.c]: `extern volatile uint32_t timeout; // global variable for timeout' This one?

**Comment by btsimonh** [comment]
note: I am developing an aversion to #defines in config.h and platformio.ini....
I removed a LOT in the last push (at least defaulted a lot if not present, so platformio.ini was cut down).
But for things like the selection of the timeout or the selection of hard stop, can we have a variable (the default of which can be a define)?  Then it can be got from flash, or set from protocol... :).
p.s. setting posn to current posn will be a very hard stop if at speed, as will setting the speed to zero :).
Also, the timeout operation should really be in one of the interrupt routines - to protect us from a crash in main where bldc is still happily ticking over....

File: inc/config.h (hunk 2 above)
Comment (AntumArk) [reply inc/config.h]: I haven't renamed the variable, I added it. `timeout` is still used at https://github.com/bipropellant/bipropellant-hoverboard-firmware/blob/master/src/control.c#L107
 I2C and with ppm and it is counts, and not time.  Same variable is used in bldc.c#L239, so for safety I do not want to touch that.

**Comment by AntumArk** [comment]
> p.s. setting posn to current posn will be a very hard stop if at speed, as will setting the speed to zero :).
Well that is why you CAN select if you want HARD stop or not :D
> Also, the timeout operation should really be in one of the interrupt routines - to protect us from a crash in main where bldc is still happily ticking over.... 
Even if it would slightly slow down the routine?

**Comment by btsimonh** [comment]
decrementing one value will not add much overhead.  Suggest you set a value to (ms * 16) each time you get some command, and decrement (only if >0) in bldc.c (DMA1_Channel1_IRQHandler()), and action in there when it hits zero.  After all, if the DMA interrupt is not running, neither are the motors...
maybe call the variable 'motor_stop_timeout_16khz' or something else obvious? and a function 'reset_motor_timer_ms( ms_value)' to set it?  Then we can call it wherever (e.g. when I have 'either_sensor_ok' true in main).... :). And setting to -1 -> no timeout?

**Comment by p-h-a-i-l** [comment]
> note: I am developing an aversion to #defines in config.h and platformio.ini....

As do I :) But I think we will grow out of them as we want more and more values stored in flash and updateable via protocol.

In regard to your proposal with using the motor interrupt as a watchdog..
What about using what you described as the watchdog for the main and handling every other monitoring in there?

**Comment by btsimonh** [comment]
>  handling every other monitoring in there?

explain?

**Comment by p-h-a-i-l** [comment]
We have multiple values to monitor. Checking the input for plausibility and timeouts, battery health, inactivity times, temperatures.

You could argue, that that some are more important than others, but this might blow up the motor interrupt. 
So instead use a watchdog (might even BE the motor interrupt) to monitor main and have everything else in there as is now.

File: inc/config.h (hunk 2 above)
Comment (p-h-a-i-l) [reply inc/config.h]: If we start to rework the timeout, we should do it properly. For all inputs

File: src/main.c (hunk 4 above)
Comment (p-h-a-i-l) [reply src/main.c]: yes, see discussion above about reworking the "old" timeout

**Comment by btsimonh** [comment]
yes, so it's true that, for example, if main stalled, we'd stop reading/acting upon temp.  And so if no other things stopped the wheels, we could over-temp.  Doing ALL these things in a 16khz high priority interrupt may be a little over agressive.
IDEA:
If we setup a low priority interrupt, can we trigger it periodically from the DMA interrupt?
i.e. we don't dedicate a timer to it; are there interrupts that we can trigger via software (see EXTI_SWIER)?
OR - WWDG - watchdog which produces an 'early' interrupt, but which can also reset the STM32.  Hence we could set it up to run for 20ms, at a few ms to go, it interrupts, we check all values, and reset the count if all is ok...

**Comment by p-h-a-i-l** [comment]
Until we have a nice and clean new system, I will push an interim solution soon. Reactivation of the "old" timeout system with minor improvements: https://github.com/bipropellant/bipropellant-hoverboard-firmware/pull/55

**Pull request closed without merging**